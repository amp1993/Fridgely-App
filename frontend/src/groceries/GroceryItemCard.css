.item-list-container {
    max-width: 600px; /* Set the maximum width */
    margin: 0 auto; /* Center the container horizontally */
  }
  
  /* Style for each todo item */
  .item-item {
    background-color: #f9f9f9;
    border-radius: 5px;
    padding: 10px;
    margin-bottom: 10px;
    display: flex;
    justify-content: space-between;
    align-items: center;
  }
  
  /* Style for todo item text */
  .item-text {
    margin: 0;
  }
  
  /* Style for todo item buttons */
  .item-buttons {
    display: flex;
    align-items: center;
  }
  
  /* Style for buttons in todo item */
  button {
    margin-left: 5px;
    cursor: pointer;
    padding: 5px 10px;
    border: none;
    border-radius: 3px;
    background-color: #007bff;
    color: #fff;
  }
  
  /* Hover effect for buttons */
  .item-buttons button:hover {
    background-color: #0056b3;
  }
  
  .item-list-container .add-item-button{
    margin-bottom: 20px;
  }