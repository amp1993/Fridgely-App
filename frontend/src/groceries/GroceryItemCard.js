import React, { useState, useEffect } from "react";
import { useNavigate } from "react-router-dom";
import FridgelyApi from "../api/api";
import "./GroceryItemCard.css"

const GroceryItemCard = ({ groceryItems, user, updateGroceryItems,  }) => {
  
  const navigate = useNavigate()
  const [sortedItems, setSortedItems] = useState([]);

  useEffect(() => {
    // Flatten the nested array of items
    const flattenedItems = groceryItems[0];
    // Sort the flattened array by categoryName
    const sorted = flattenedItems.sort((a, b) => a.categoryName.localeCompare(b.categoryName));
    // Set the sorted items in state
    setSortedItems(sorted)
  }, [groceryItems]);



  // Group the sorted items by categoryName
  const groupedItems = sortedItems.reduce((acc, item) => {
    const category = item.categoryName;
    if (!acc[category]) {
      acc[category] = [];
    }
    acc[category].push(item);
    return acc;
  }, {});


  // Handle quantity increase
  const handleIncrease = async (productName) => {
    // Find the current item
    const currentItem = groceryItems[0].find(item => item.productName === productName);
    if (!currentItem) {
      console.error("Item not found");
      return;
    }

    // Increment the quantity
    const newQuantity = parseInt(currentItem.quantityInGroceryList,10) + 1;

    try {
      // Update the item in the backend
      await FridgelyApi.updateQtyGroceryItem({ quantityInGroceryList: newQuantity }, productName, user);
      // Fetch updated data and update state
      const fetchUpdatedData = await FridgelyApi.getGroceryItems(user);
      updateGroceryItems(fetchUpdatedData);

      

    } catch (error) {
      console.error("Error updating item:", error);
    }
  };

  // Handle quantity decrease
  const handleDecrease = async (productName) => {
    // Find the current item
    const currentItem = groceryItems[0].find(item => item.productName === productName);
    if (!currentItem) {
      console.error("Item not found");
      return;
    }
    // Decrement the quantity but ensure it doesn't go below 0
    const newQuantity = parseInt(Math.max(currentItem.quantityInGroceryList - 1, 0));
    try {
      await FridgelyApi.updateQtyGroceryItem({ quantityInGroceryList: newQuantity }, productName, user);
      // Fetch updated data and update state
      const fetchUpdatedData = await FridgelyApi.getGroceryItems(user);
      updateGroceryItems(fetchUpdatedData);


    } catch (error) {
      console.error("Error updating item:", error);
    }
  };

  // Handle item deletion
  const handleDelete = async (productName) => {
    try {
      await FridgelyApi.removeItemFromGroceryList(productName, user);
      // Fetch updated data and update state
      const fetchUpdatedData = await FridgelyApi.getGroceryItems(user);
      updateGroceryItems(fetchUpdatedData);
      

    } catch (error) {
      console.error("Error deleting item:", error);
    }
  };

  // Move item to fridge
  const moveToFridge = async (productName, quantity) =>{
    try{
      //Update grocery item and quantity to fridge
      await FridgelyApi.updateGroceryItem({quantityInFridge:quantity} ,productName,user);
      // Fetch updated grocery list data and update state
      const fetchUpdatedGroceryList = await FridgelyApi.getGroceryItems(user);

      updateGroceryItems(fetchUpdatedGroceryList);  
    } catch (error){
      console.error("Error deleting item:", error);

    }
  };
  // Navigate to new item form
  const addItem = ()=>{
    navigate(`/groceries/${user}`)
  }
  return (
    <div className="item-list-container">
          
        <button className="add-item-button" onClick={()=>addItem()}>Add an Item</button>

      {Object.keys(groupedItems).map(category => (
        <div key={category}>
          <h3>{category}</h3>
          <div className="item-list">
            {groupedItems[category].map(groceryItem => (
              <li className="item-item" key={groceryItem.id}>
                <p className="item-text"><strong>{groceryItem.productName}</strong></p>
                <div className="item-buttons">
                  <button onClick={() => handleDecrease(groceryItem.productName)}>←</button>
                  <p style={{ margin: '0 10px' }}>
                  {groceryItem.quantityInGroceryList} {groceryItem.unitOfMeasure}
                </p>
                  <button onClick={() => handleIncrease(groceryItem.productName)}>→</button>
                  <button onClick={() => handleDelete(groceryItem.productName)}>Delete</button>
                  <button className="item-purchased-button" onClick={() => moveToFridge(groceryItem.productName, groceryItem.quantityInGroceryList)}>✓</button>
                </div>
               
              </li>
            ))}
          </div>
        </div>
      ))}

    </div>
  );
  
  
  
};

export default GroceryItemCard;
