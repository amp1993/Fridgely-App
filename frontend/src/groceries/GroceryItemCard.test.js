import React from 'react';
import { render } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import GroceryItemCard from './GroceryItemCard';

const mockGroceryItems = [
  { id: 1, productName: 'Milk', categoryName: 'Dairy', quantityInGroceryList: 2, unitOfMeasure: 'Liters' },
  { id: 2, productName: 'Bread', categoryName: 'Bakery', quantityInGroceryList: 1, unitOfMeasure: 'Loaf' },
];

it('renders without crashing', function() {
  render(
    <BrowserRouter>
      <GroceryItemCard groceryItems={[mockGroceryItems]} user="testuser" updateGroceryItems={() => {}} />
    </BrowserRouter>
  );
});

it('matches snapshot', function() {
  const { asFragment } = render(
    <BrowserRouter>
      <GroceryItemCard groceryItems={[mockGroceryItems]} user="testuser" updateGroceryItems={() => {}} />
    </BrowserRouter>
  );
  expect(asFragment()).toMatchSnapshot();
});

it('displays grocery items correctly', function() {
  const { getByText } = render(
    <BrowserRouter>
      <GroceryItemCard groceryItems={[mockGroceryItems]} user="testuser" updateGroceryItems={() => {}} />
    </BrowserRouter>
  );

  expect(getByText('Milk')).toBeInTheDocument();
  expect(getByText('Bread')).toBeInTheDocument();
});
