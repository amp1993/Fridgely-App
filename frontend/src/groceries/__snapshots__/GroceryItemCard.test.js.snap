// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="item-list-container"
  >
    <button
      class="add-item-button"
    >
      Add an Item
    </button>
    <div>
      <h3>
        Bakery
      </h3>
      <div
        class="item-list"
      >
        <li
          class="item-item"
        >
          <p
            class="item-text"
          >
            <strong>
              Bread
            </strong>
          </p>
          <div
            class="item-buttons"
          >
            <button>
              ←
            </button>
            <p
              style="margin: 0px 10px;"
            >
              1 Loaf
            </p>
            <button>
              →
            </button>
            <button>
              Delete
            </button>
            <button
              class="item-purchased-button"
            >
              ✓
            </button>
          </div>
        </li>
      </div>
    </div>
    <div>
      <h3>
        Dairy
      </h3>
      <div
        class="item-list"
      >
        <li
          class="item-item"
        >
          <p
            class="item-text"
          >
            <strong>
              Milk
            </strong>
          </p>
          <div
            class="item-buttons"
          >
            <button>
              ←
            </button>
            <p
              style="margin: 0px 10px;"
            >
              2 Liters
            </p>
            <button>
              →
            </button>
            <button>
              Delete
            </button>
            <button
              class="item-purchased-button"
            >
              ✓
            </button>
          </div>
        </li>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
